middleware/hosts/hosts.go

```diff
@@ -0,0 +1,178 @@
+package hosts
+
+import (
+	"errors"
+	"net"
+
+	"golang.org/x/net/context"
+
+	"github.com/coredns/coredns/middleware"
+	"github.com/coredns/coredns/middleware/pkg/dnsutil"
+	"github.com/coredns/coredns/request"
+	"github.com/miekg/dns"
+)
+
+// Hosts is the middleware handler
+type Hosts struct {
+	Next middleware.Handler
+	*Hostsfile
+
+	Origins     []string
+	Fallthrough bool
+}
+
+// ServeDNS implements the middleware.Handle interface.
+func (h Hosts) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+	if state.QClass() != dns.ClassINET {
+		return dns.RcodeServerFailure, middleware.Error(h.Name(), errors.New("can only deal with ClassINET"))
+	}
+	qname := state.Name()
+
+	answers := []dns.RR{}
+
+	// Precheck with the origins, i.e. are we allowed to looks here.
+	if h.Origins != nil {
+		zone := middleware.Zones(h.Origins).Matches(qname)
+		if zone == "" {
+			// PTR zones don't need to be specified in Origins
+			if state.Type() != "PTR" {
+				// If this doesn't match we need to fall through regardless of h.Fallthrough
+				return middleware.NextOrFailure(h.Name(), h.Next, ctx, w, r)
+			}
+		}
+	}
+
+	zone := middleware.Zones(h.Names()).Matches(qname)
+	if zone == "" {
+		if state.Type() != "PTR" {
+			return h.NextOrFailure(h.Name(), h.Next, ctx, w, r)
+		}
+
+		// Request is a PTR
+		zone = state.Name()
+		names := h.LookupStaticAddr(dnsutil.ExtractAddressFromReverse(zone))
+		if len(names) == 0 {
+			// If this doesn't match we need to fall through regardless of h.Fallthrough
+			return middleware.NextOrFailure(h.Name(), h.Next, ctx, w, r)
+		}
+		answers = h.ptr(zone, names)
+	}
+
+	if zone != "" {
+		ips := h.LookupStaticHost(zone)
```

Comment:
Wouldn't it be better to let `LookupStatisHost` returns `net.IP`s? Now we keep on parsing the same string over and over again in `ipv4Filter` and `ipv6Filter`...